### PR TITLE
Move DPRNN to padertorch

### PR DESCRIPTION
**Pull request by thequilo**
This PR moves the DPRNN implementation from padercontrib to padertorch. It still needs a little bit of cleanup and testing.

Things that are not that clear yet:
 - where to put the chunking: contrib/neumann or data?
 - put the training script in contrib or examples?

### CODE REVIEW

**Comment by jensheit** [comment]
In my opinion at least the train and evaluate script should be moved to contrib.examples.dual_path_rnn. The same is true for the README.md. One could argue that the chunking could stay in the current path.

**Comment by jensheit** [comment]
In my opinion at least the train and evaluate script should be moved to contrib.examples.dual_path_rnn. The same is true for the README.md. One could argue that the chunking could stay in the current path.

**Comment by jensheit** [comment]
We should think of a better name for the model and modules, perhaps we could create an additional directory modules/tasnet/encoder modules/tasnet/decoder and change the model to dprnn_source_separation or create additonal sub directories here as well.

File: padertorch/models/dual_path_rnn_source_separator.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+    return {
+        'relu': torch.nn.ReLU(),
+        'sigmoid': torch.nn.Sigmoid(),
+        'softmax': torch.nn.Softmax(dim=speaker_dim),
+        None: identity,
+        'None': identity,
+        'identity': identity,
+    }[nonlinearity]
+
```
Comment (jensheit) [inline padertorch/models/dual_path_rnn_source_separator.py:26]: For get_nonlinearities you should use the pt.ops.mapping functions

File: padertorch/models/dual_path_rnn_source_separator.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        return {k: torch.mean(torch.stack(v)) for k, v in losses.items()}
+
+    def review(self, inputs, outputs):
+        # Report audios
+        audios = {
+            'observation': pt.summary.audio(
+                signal=inputs['y'][0], sampling_rate=8000
+            ),
+        }
+
```
Comment (jensheit) [inline padertorch/models/dual_path_rnn_source_separator.py:261]: pt.summary.audio does not exist, probably you did not commit your change to the __init__.py

Comment (jensheit) [inline padertorch/models/dual_path_rnn_source_separator.py:261]: pt.summary.audio does not exist, probably forgot to commit a change to all in tbx_utils

File: padertorch/ops/losses/__init__.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
 from . import loss
+from . import time_domain_losses
```
Comment (jensheit) [inline padertorch/ops/losses/__init__.py:2]: if we add a time_domain_losses.py we probably have to rename the old loss or/and split it into multiple files

File: padertorch/models/dual_path_rnn_source_separator.py (hunk 2 above)
Comment (thequilo) [reply padertorch/models/dual_path_rnn_source_separator.py:261]: `pt.summary.audio` is a quite old change. It exists in `padertorch.summary.tbx_utils` and is imported to `pt.summary`. If not, something is very wrong with the git repository.

Comment (jensheit) [reply padertorch/models/dual_path_rnn_source_separator.py:261]: I have pulled your request and I had to add audio to the __all__[] command at the begining of the tbx_utils.py file to get the script to work

File: padertorch/ops/losses/__init__.py (hunk 3 above)
Comment (thequilo) [reply padertorch/ops/losses/__init__.py:2]: I'm not sure what the best naming is here. I first put these losses into `loss`, but the file becomes pretty large then.

File: padertorch/models/dual_path_rnn_source_separator.py (hunk 2 above)
Comment (thequilo) [reply padertorch/models/dual_path_rnn_source_separator.py:261]: Hmm, works for me without that. I'll add it to `__all__`

File: padertorch/contrib/examples/tasnet/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+| `with log_mse`  | 16.7  | 17.0  |
+
+Training
+--------
+
+A storage root must be set with `export STORAGE=/path/to/your/storage`.
+After installing `padercontrib`, a training can be started with
+
+```bash
+$ python -m padercontrib.pytorch.examples.dual_path_rnn.train
```
Comment (boeddeker) [inline padertorch/contrib/examples/tasnet/README.md:24]: Can you change this to `padertorch` instead of `padercontrib`

**Comment by thequilo** [comment]
The DPRNN-TasNet example should work fine now except for the dependency on `padercontrib.database.wsj0_mix.WSJ0_2Mix_8k`. What should we do about that?

**Comment by boeddeker** [comment]
Can you reduce the dependency, so that you only depend on the json that is used in `padercontrib.database.wsj0_mix.WSJ0_2Mix_8k` and not the class?
Most of the properties are already public in the super class: `lazy_dataset_database.JsonDatabase`

**Comment by thequilo** [comment]
I removed the import from padercontrib. The script now directly uses `lazy_dataset.database.JsonDatabase` and the path to the JSON has to be set in the configuration.

File: padertorch/contrib/examples/tasnet/train.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    iterator = db.get_dataset(dataset)
+
+    if iterator_slice is not None:
+        iterator = iterator[iterator_slice]
+
+    iterator = (
+        iterator
+        .map(pre_batch_transform)
+        .map(RandomChunkSingle(chunk_size, chunk_keys=('y', 's'), axis=-1))
+        .shuffle(reshuffle=False)
```
Comment (boeddeker) [inline padertorch/contrib/examples/tasnet/train.py:157]: Is there a reason, why you disabled reshuffle? Here it should be supported.

File: padertorch/contrib/examples/tasnet/train.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        prepare_iterable_captured(db, validate_dataset),
+    )
+
+    # Register hooks and start the actual training
+    trainer.register_validation_hook(
+        prepare_iterable_captured(db, validate_dataset)
+    )
+
+    # Learning rate scheduler
+    trainer.register_hook(pt.train.hooks.LRSchedulerHook(
```
Comment (boeddeker) [inline padertorch/contrib/examples/tasnet/train.py:272]: I think @jensheit removed the `pt.train.hooks.LRSchedulerHook` in favor of the BackOffValidationHook

Comment (jensheit) [reply padertorch/contrib/examples/tasnet/train.py:272]: Yes, and the BackOffValidationHook should be added with register_validation_hook()

Comment (thequilo) [reply padertorch/contrib/examples/tasnet/train.py:272]: But those two do different things. This one is as close to the original paper as possible. I don't know how they compare in terms of training behavior, but the backoff validation hook might be preferable

File: padertorch/contrib/examples/tasnet/train.py (hunk 5 above)
Comment (thequilo) [reply padertorch/contrib/examples/tasnet/train.py:157]: No, there's no reason for that. I think this was copied from the pit example

Comment (boeddeker) [reply padertorch/contrib/examples/tasnet/train.py:157]: Can you change this to `reshuffle=True`? The PIT example was written, before https://github.com/fgnt/lazy_dataset/pull/2 was merged.

Comment (thequilo) [reply padertorch/contrib/examples/tasnet/train.py]: Done!

File: padertorch/contrib/examples/tasnet/train.py (hunk 6 above)
Comment (boeddeker) [reply padertorch/contrib/examples/tasnet/train.py:272]: The `pt.train.hooks.LRSchedulerHook` is still available.
In #11 only LRSchedulers in `pt.train.optimizer` were deleted, `pt.train.hooks.LRSchedulerHook` is an alternative to the LRSchedulers in `pt.train.optimizer` and is now the recommended one, when no backoff is used/needed.
